chainlink-modsec/libmodsec/internal/executor/main.go

```diff
@@ -0,0 +1,117 @@
+package executor
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+func (e *Executor) run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	var wg sync.WaitGroup
+
+	// 1. We read messages from all sources for any new incoming work.
+	// Immediately we perform checks to see if it's our turn to execute the message, and skip additional processing if not.
+	for _, reader := range e.MessageSources {
+		go func(r MessageReader) {
+			wg.Add(1)
+			defer wg.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					fmt.Println("Executor shutting down message reader...")
+					return
+				case msg := <-reader.SubscribeMessages(ctx):
+					if e.isMyTurn(msg) {
+						continue
+					}
+					if e.Transmitters[msg.Header.DestChainSelector] == nil {
+						fmt.Printf("No chain writer for destination chain selector %s, skipping message %s\n",
+							msg.Header.DestChainSelector, msg.Header.MessageID)
+						continue
+					}
+					e.messageCh <- msg
+				}
+			}
+		}(reader)
+	}
+
+	transmitterPayloadCh := make(chan TransmitterPayload)
+
+	// 2. Now that we know we have messages that we are assigned to execute, try to get attestations for them.
+	//
+	go func() {
+		wg.Add(1)
+		defer wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				fmt.Println("Executor shutting down attestation poller...")
+				return
+			case msg := <-e.messageCh:
+
+				// TODO: We need some abstraction here to loop and handle retries when attestations are not yet ready
+				attestations, err := e.AttestationReader.GetAttestations(ctx, msg)
+				if err != nil {
+					// handle error, log it, etc.
+					continue
+				}
```

Comment:
See "Verifier Watcher", "Per Verifier Launcher", and "Attestation Aggregator" in the design diagram I shared, I believe they play this role.